src/NServiceBus.Core/Recoverability/Faults/FaultsStatusStorage.cs

```diff
@@ -0,0 +1,27 @@
+﻿using System;
+
+namespace NServiceBus.Recoverability.Faults
+{
+    using System.Collections.Concurrent;
+
+    class FaultsStatusStorage
+    {
+        public void AddException(string messageId, Exception exception)
+        {
+            failedMessages.AddOrUpdate(messageId, exception, (m, e) => e);
+        }
+
+        public bool TryGetException(string messageId, out Exception exception)
+        {
+            return failedMessages.TryGetValue(messageId, out exception);
+        }
+
+        public void ClearExceptions(string messageId)
```

Comment:
I think you have the same memory leak issue we fixed on the FLR storage. In case of competing consumers on the same queue it's possible to add an exception to this storage and never clear it again?